wordpress-mobile/WordPress-iOS
# WPWebViewController: Revamp Mark II

#### Description:
This Pull Request revisits all of the current interactions with WPWebViewController, and converts (almost all of them) into Modal interactions. Exceptions outlined below.

Also, several UI enhancements have been introduced: the bottom toolbar has been brought back, and whenever the WebViewController is displayed modally, it'll skin the NavigationBar with a different style.

#### Notes:
- **StatsWebViewController** _might_ push a **WebViewController** instance, when the user taps over an external link. We'll deal with those interactions as we always did [as a navigationController push].
- **AboutViewController** can display  **WebViewController** instances. Since About can be displayed either modally or "pushed". We'll avoid presenting the webViewController modally, if we're already in a modal scenario.
- Post previews are actually handled by **PostPreviewViewController**, so this PR doesn't affect that functionality.

#### Review:
Needs Review: @aerych + @astralbodies 

Eric + Aaron: may i bother you guys with a review?. Code delta isn't huge, but it'll probably take a while to test everything. Maybe 7 test cases / 7 test cases, and raffle the last one?

(Thanks in advance!!)



Closes #3763

--

#### Test 1: Login
1. Fresh install WPiOS
2. Tap over the 'Add Self-Hosted Site' label
3. Enter invalid credentials, and hit Sign In
4. Once you've gotthe 'Sorry, we can't log you in' screen, tap over the bottom left icon that reads 'Need Help?'

**Expected:** Modal WebViewController onscreen

--

#### Test 2: Create Account
1. Launch WPiOS
2. Tap over Create Account
3. Tap the Terms of Service label

**Expected:** Modal WebViewController onscreen

--

#### Test 3: About (Logged Off)
1. Fresh install WPiOS
2. Tap over the top right (?) icon
3. Tap over About, and pick any of the links

**Expected:** WebViewController onscreen, with the regular NavBar style

--

#### Test 4: About (Logged In)
1. Log into WPiOS
2. Tap over the Me tab, and hit the *Help & Support* row
3. Tap over About, and pick any of the links

**Expected:** Modal WebViewController onscreen

--

#### Test 5: View Site
1. Launch WPiOS and log into your account
2. Tap over My Sites, and pick a random blog
3. Tap over the *View Site* label

**Expected:** Modal WebViewController onscreen

--

#### Test 6: Reader Post Image
1. Launch WPiOS and log into your account
2. Tap over the Reader and pick any article
3. Tap over an image that's not PNG | JPG | GIF

**Expected:** Modal WebViewController onscreen
**Note:** This is way easier to test if you just hack *WPImageViewController*'s isUrlSupported method, to always return NO.

--

#### Test 7: Reader Comment Image
1. Launch WPiOS and log into your account
2. Tap over the Reader, pick any article, and open its comments
3. Tap over an image that's not PNG | JPG | GIF

**Expected:** Modal WebViewController onscreen
**Note:* This is way easier to test if you just hack *WPImageViewController*'s isUrlSupported method, to always return NO.

--

#### Test 8: Reader Post Link
1. Launch WPiOS and log into your account
2. Tap over the Reader and pick any article
3. Tap over any external links

**Expected:** Modal WebViewController onscreen

--

#### Test 9: Reader Comment Link
1. Launch WPiOS and log into your account
2. Tap over the Reader, pick any article, and open its comments
3. Tap over any external links

**Expected:** Modal WebViewController onscreen

--

#### Test 10: Reader Browse Site
1. Launch WPiOS and log into your account
2. Tap over the Reader, and tap over the top left icon of any article (Blavatar)
3. Tap again over the Blavatar

**Expected:** Modal WebViewController onscreen

--

#### Test 11: Notifications Jetpack
1. Launch WPiOS and log into a self hosted blog, with no Jetpack
2. Tap over the Notifications tab
3. Tap over the "Learn More" button

**Expected:** Modal WebViewController onscreen
**Note:** This is easier to test with a WordPress.com account with no notifications. Just hack the *NotificationsViewController*'s showsJetpackMessage method to always return YES.

--

#### Test 12: Notification Details
1. Launch WPiOS and log into your account
2. Tap over the Notifications tab, and tap over any notification that contains a web link (ie. badge / comment)
3. Tap over the link

**Expected:** Modal WebViewController onscreen

--

#### Test 13: Comments
1. Launch WPiOS and log into your account
2. Tap over My Sites, and pick a random blog
3. Tap over 'Comments', and pick a comment that contains a web link
4. Tap over the link

**Expected:** Modal WebViewController onscreen

--

#### Test 14: Edit Site
1. Log into a self hosted blog
2. Tap over My Sites, and pick up the blog
3. Force an error 405 (easiest way: hack the *validationDidFail* method in EditSiteViewController, change the password onscreen, and hit save).

**Expected:** Modal WebViewController onscreen

-- 

#### Test 15: Install Jetpack
1. Log into a self hosted blog, that doesn't have Jetpack installed
2. Tap over My Sites > Stats
3. Tap over the 'Install Jetpack' button

**Expected:** Modal WebViewController onscreen
